service/src/java/org/apache/hive/service/servlet/OTELExporter.java

```diff
@@ -0,0 +1,229 @@
+      } else {
+        // In case of live queries being seen for first time
+        Span rootspan = tracer.spanBuilder(queryID + " - live")
+                .startSpan();
+        List<String> completedTasks = new ArrayList<>();
+        for (QueryDisplay.TaskDisplay task : lQuery.getQueryDisplay().getTaskDisplays()) {
+          if (task.getReturnValue() != null) {
+            completedTasks.add(task.getTaskId());
+            Context parentContext = Context.current().with(rootspan);
+            Span currSpan = tracer.spanBuilder(queryID + " - " + task.getTaskId() + " - live").setParent(parentContext).setAllAttributes(addTaskAttributes(task))
+                    .setStartTimestamp(task.getBeginTime(), TimeUnit.MILLISECONDS).startSpan();
+
+            while (task.getEndTime() == null) {
+              try {
+                Thread.sleep(500);
+              } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+              }
+            }
+            currSpan.end(task.getEndTime(), TimeUnit.MILLISECONDS);
+          }
+        }
+
+        if(rootspan != null){
+          queryIdToSpanMap.put(queryID,rootspan);
+          queryIdToTasksMap.put(queryID,completedTasks);
+        }
+
+      }
+    }
+
+
+    for (QueryInfo hQuery : historicalQueries) {
+
+      //For queries that were live till last loop but have ended before start of this loop
+      if(queryIdToSpanMap.containsKey(hQuery.getQueryDisplay().getQueryId())){
+        String hQueryId = hQuery.getQueryDisplay().getQueryId();
+        Span rootspan = queryIdToSpanMap.get(hQueryId);
+        for (QueryDisplay.TaskDisplay task : hQuery.getQueryDisplay().getTaskDisplays()) {
+          if(queryIdToTasksMap.get(hQueryId) == null || !queryIdToTasksMap.get(hQueryId).contains(task.getTaskId())){
+            queryIdToTasksMap.get(hQueryId).add(task.getTaskId());
+            Context parentContext = Context.current().with(rootspan);
+            Span currSpan = tracer.spanBuilder(hQueryId+ " - " + task.getTaskId() + " - completed").setParent(parentContext).setAllAttributes(addTaskAttributes(task))
+                    .setStartTimestamp(task.getBeginTime(), TimeUnit.MILLISECONDS).startSpan();
+            currSpan.end(task.getEndTime(), TimeUnit.MILLISECONDS);
+          }
+        }
+
+        //Update the rootSpan name & attributes before ending it
+        queryIdToSpanMap.get(hQueryId).updateName(hQueryId + " - completed");
+        queryIdToSpanMap.get(hQueryId).setAllAttributes(addQueryAttributes(hQuery));
+        queryIdToSpanMap.get(hQueryId).end();
+
+        queryIdToSpanMap.remove(hQueryId);
+        queryIdToTasksMap.remove(hQueryId);
+
+        historicalQueryId.add(hQueryId);
+      }
+
+      //For queries that were already either before OTEL service started or in between OTEL loops
+      if (!historicalQueryId.contains(hQuery.getQueryDisplay().getQueryId())) {
+        historicalQueryId.add(hQuery.getQueryDisplay().getQueryId());
```

Comment:
It is a Set, you don't need to do contains then add, you can directly add and check the return value, if it is true, means the element wasn't there, if false, means the element was already there